### ISSUE TITLE
useInfiniteQuery: hasNextPage does not work with custom select function

### ISSUE THREAD

**Issue by danbeneventano** [tannerlinsley/react-query#1537]
```
useInfiniteQuery(['posts', variables], fetchPosts, {
    getNextPageParam: (lastPage, pages) => lastPage.nextPage
    select: data => data.pages.flatMap(page => page.posts)
  })
```

`hasNextPage` always returns `undefined` when I specify a `select` function. When I remove the `select` function, `hasNextPage` works as expected.

**Comment by boschni** [comment]
When using `select` you need to return the original data structure which contains `pages` and `pageParams`:
```js
useInfiniteQuery(['posts', variables], fetchPosts, {
  getNextPageParam: lastPage => lastPage.nextPage,
  select: data => ({ ...data, pages: data.pages.flatMap(page => page.posts) })
})
```
There was a bug though because the selected data was passed into `getNextPageParam`. This has been fixed in `3.5.6`.

**Comment by TkDodo** [comment]
@boschni will this not void the possibility to subscribe to parts of the state via `select` if we have to create a new object to return?
